components/binding/internal/target/handler.go

```diff
@@ -1,134 +1,223 @@
+		storage: storage,
+	}
 }
 
 func (h *Handler) AddLabel(b *v1alpha1.Binding) error {
-	ctx := context.Background()
-
-	deployment, err := h.getDeployment(ctx, b)
+	resourceData, err := h.storage.Get(storage.Kind(b.Spec.Target.Kind))
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "while getting Kind %s from storage", b.Spec.Target.Kind)
 	}
-
-	if deployment.Spec.Template.ObjectMeta.Labels == nil {
-		deployment.Spec.Template.ObjectMeta.Labels = make(map[string]string, 0)
+	resource, err := h.getResource(b, resourceData)
+	if err != nil {
+		return errors.Wrapf(err, "while getting resource for Binding %s", b.Name)
 	}
-	deployment.Spec.Template.ObjectMeta.Labels[h.labelKey(b)] = uuid.New().String()
-
-	if err := h.updateDeployment(ctx, deployment); err != nil {
-		return err
+	labelsToApply := map[string]string{h.labelKey(b): uuid.New().String()}
+	if err := h.ensureLabelsAreApplied(resource, labelsToApply, resourceData.LabelFields); err != nil {
+		return errors.Wrap(err, "while ensuring labels are applied")
+	}
+	err = h.updateResource(resource, resourceData)
+	if err != nil {
+		return errors.Wrap(err, "while updating resource")
 	}
 
 	return nil
 }
 
 func (h *Handler) LabelExist(b *v1alpha1.Binding) (bool, error) {
-	ctx := context.Background()
-
-	deployment, err := h.getDeployment(ctx, b)
+	resourceData, err := h.storage.Get(storage.Kind(b.Spec.Target.Kind))
 	if err != nil {
-		return false, err
+		return false, errors.Wrapf(err, "while getting Kind %s from storage", b.Spec.Target.Kind)
+	}
+	resource, err := h.getResource(b, resourceData)
+	if err != nil {
+		return false, errors.Wrapf(err, "while getting resource for Binding %s", b.Name)
 	}
 
-	for key, _ := range deployment.Spec.Template.ObjectMeta.Labels {
+	resourceLabels, err := h.getResourceLabels(resource, resourceData.LabelFields)
+	if err != nil {
+		return false, errors.Wrapf(err, "while getting injected labels for Binding %s", b.Name)
+	}
+	for key, _ := range resourceLabels {
 		if key == h.labelKey(b) {
 			return true, nil
 		}
-	}
 
+	}
 	return false, nil
 }
 
 func (h *Handler) RemoveOldAddNewLabel(b *v1alpha1.Binding) error {
-	ctx := context.Background()
-
-	deployment, err := h.getDeployment(ctx, b)
+	resourceData, err := h.storage.Get(storage.Kind(b.Spec.Target.Kind))
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "while getting Kind %s from storage", b.Spec.Target.Kind)
+	}
+	resource, err := h.getResource(b, resourceData)
+	if err != nil {
+		return errors.Wrapf(err, "while getting resource for Binding %s", b.Name)
 	}
 
-	for key, _ := range deployment.Spec.Template.ObjectMeta.Labels {
+	resourceLabels, err := h.getResourceLabels(resource, resourceData.LabelFields)
+	if err != nil {
+		return errors.Wrapf(err, "while getting injected labels for Binding %s", b.Name)
+	}
+	for key, _ := range resourceLabels {
 		if key == h.labelKey(b) {
-			deployment.Spec.Template.ObjectMeta.Labels[h.labelKey(b)] = uuid.New().String()
```

Comment:
I don't understand this part: you fetch all labels, then iterate for keys, and if a label exists you `ensureLabelsAreApplied` ?